File: src/gestorAplicacion/personas/Empleado.java
package gestorAplicacion.personas;

import java.time.LocalDate;
import java.util.ArrayList;

import gestorAplicacion.general.DiaSemana;
import gestorAplicacion.general.Inventario;
import gestorAplicacion.ventas.Servicio;

/**
 * Esta clase extiende de persona y se encarga de manejar los empleados de la aplicacion
 * los cuales se usan para atender a los servicios, la creacion de facturas y manejar su 
 * proceso de contratacion.
 * 
 * @author Mateo Alvarez Lebrum
 * @author Alejandro Alvarez Botero
 * @author Miguel Angel Barrera Bustamante
 * @author Alejandra Barrientos Grisales
 */
public class Empleado extends Persona {
	private static final long serialVersionUID = 1L;

	private Contrato contrato;
	private String cargo;
	private Servicio servicio;
	private ArrayList<DiaSemana> diasLaborales = new ArrayList<DiaSemana>();

	/**
	 * Este constructor se usa para la construccion generica desde 0 de los empleados, es decir,
	 * cuando no existe una persona previamente relacionada.
	 * 
	 * @param nombre
	 * @param telefono
	 * @param email
	 * @param identificacion
	 * @param tipoDeIdentificacion
	 * @param sexo
	 * @param contrato
	 * @param cargo
	 * @param servicio
	 * @param diasLaborales
	 */
	public Empleado(String nombre, String telefono, String email, String identificacion,
			TipoDocumento tipoDeIdentificacion, Sexo sexo, Contrato contrato, String cargo, Servicio servicio,
			ArrayList<DiaSemana> diasLaborales) {
		super(nombre, telefono, email, identificacion, tipoDeIdentificacion, sexo);
		this.contrato = contrato;
		this.cargo = cargo;
		this.servicio = servicio;
		this.diasLaborales = diasLaborales;

	}

	/**
	 * Este constructor se usa para la contratacion para cuando se crea un empleado
	 * a partir de una persona ya existente. En este se elimina la persona que existia previamente
	 * ya que por el comportamiento de la herencia, quedaria la persona duplicada. De igual manera,
	 * como no hay nada asociado a la persona, no hay problema con la consistencia de datos.
	 * 
	 * @param persona
	 * @param contrato
	 * @param cargo
	 * @param servicio
	 * @param diasLaborales
	 */
	public Empleado(Persona persona, Contrato contrato, String cargo, Servicio servicio,
			ArrayList<DiaSemana> diasLaborales) {
		super(persona.nombre, persona.telefono, persona.email, persona.identificacion, persona.tipoDeIdentificacion,
				persona.sexo);
		Inventario.eliminarPersona(persona);
		this.contrato = contrato;
		this.cargo = cargo;
		this.servicio = servicio;
		this.diasLaborales = diasLaborales;

	}

	public Contrato getContrato() {
		return contrato;
	}

	public void setContrato(Contrato contrato) {
		this.contrato = contrato;
	}

	public String getCargo() {
		return cargo;
	}

	public void setCargo(String cargo) {
		this.cargo = cargo;
	}

	public Servicio getServicio() {
		return servicio;
	}

	public void setServicio(Servicio servicio) {
		this.servicio = servicio;
	}

	public ArrayList<DiaSemana> getDiasLaborales() {
		return diasLaborales;
	}

	public void setDiasLaborales(ArrayList<DiaSemana> diasLaborales) {
		this.diasLaborales = diasLaborales;
	}

	/**
	 * Verifica que el empleado que este activo, es decir, tenga un contrato vigente. Adicionalmente
	 * verifica si esta disponible en el dia solicitado, esto se hace segun el dia de la semana correspondiente.
	 * 
	 * @param servicio
	 * @param fechaSolicitud
	 * @return disponible
	 */
	public boolean consultarDisponibilidadEmpleado(Servicio servicio, LocalDate fechaSolicitud) {
		boolean disponible = false;
		if (this.servicio == servicio && this.isActivo(fechaSolicitud)) {
			for (DiaSemana i : diasLaborales) {
				if (i.ordinalDia - 1 == (fechaSolicitud.getDayOfWeek().ordinal())) {
					disponible = true;
				}
			}
		}
		return disponible;
	}

	/**
	 * Consulta si el contrato esta vigente actualmente
	 * @return boolean
	 */
	public boolean isActivo() {
		return contrato.consultarVigencia(LocalDate.now());
	}
	
	/**
	 * Consulta si el contrato esta vigente en la fecha especificada
	 * @return boolean
	 */
	public boolean isActivo(LocalDate fecha) {
		return contrato.consultarVigencia(fecha);
	}
	
	
	/**
	 * Devuelve un texto con informacion del empleado dependiendo si tiene contrato vigente o no.
	 * @return informacion
	 */
	public String mostrarInformacion() {
		String informacion = "";
		if (this.isActivo()) {
			informacion = "Soy el empleado con contrato vigente " + nombre + " con numero de identificacion: "
					+ identificacion;
		} else {
			informacion = "Soy el empleado sin contrato vigente " + nombre + " con numero de identificacion: "
					+ identificacion;
		}
		return informacion;
	}

	/**
	 * Establece el dia actual como fecha final del contrato para realizar su despido
	 */
	public void despedir() {
		LocalDate hoy = LocalDate.now();
		this.contrato.setFechaFin(hoy);
	}
	/**
	 * Establece una nueva fecha fin del contrato para alargar su vigencia
	 */
	public void renovarContrato(LocalDate fechaFin) {
		if (fechaFin.isAfter(contrato.getFechaFin())) {
			this.contrato.setFechaFin(fechaFin);
		}

	}
}

File: python/src/gestorAplicacion/personas/Empleado.py
from datetime import datetime

from gestorAplicacion.general.DiaSemana import DiaSemana  
from gestorAplicacion.ventas.Servicio import Servicio 
from gestorAplicacion.personas.Persona import Persona
from gestorAplicacion.personas.Contrato import Contrato
from gestorAplicacion.personas.TipoDocumento import TipoDocumento 
from gestorAplicacion.personas.Sexo import Sexo 

'''
# @author Alejandro Alvarez
# @author Alejandra Barrientos
# @author Mateo Alvarez
# @author Miguel Barrera'''

class Empleado(Persona):
    """
    Esta clase extiende de persona y se encarga de manejar los empleados de la
    aplicacion los cuales se usan para atender a los servicios, la creacion de
    facturas y manejar su proceso de contratacion.
    """

    def __init__(self, nombre : str, telefono : int, email : str, identificacion : int, tipoDeIdentificacion : TipoDocumento, sexo : Sexo, contrato : Contrato, cargo : str, servicio : Servicio, diasLaborales : 'list[DiaSemana]'):
        super().__init__(nombre, telefono, email, identificacion, tipoDeIdentificacion, sexo)
        self._contrato = contrato
        self._cargo = cargo
        self._servicio = servicio
        self._diasLaborales = diasLaborales


    def getContrato(self) -> Contrato:
        return self._contrato

    def setContrato(self, contrato) -> None:
        self._contrato = contrato

    def getCargo(self) -> str:
        return self._cargo

    def setCargo(self, cargo : str) -> None:
        self._cargo = cargo

    def getServicio(self) -> Servicio:
        return self._servicio

    def setServicio(self, servicio : Servicio) -> None:
        self._servicio = servicio

    def getDiasLaborales(self) -> 'list[DiaSemana]':
        return self._diasLaborales

    def setDiasLaborales(self, diasLaborales : 'list[DiaSemana]'):
        self._diasLaborales = diasLaborales

    def consultarDisponibilidadEmpleado(self, servicio : Servicio, fechaSolicitud : datetime) -> bool:
        """Verifica que el empleado que este activo, es decir, tenga un contrato
        vigente. Adicionalmente verifica si esta disponible en el dia solicitado,
        esto se hace segun el dia de la semana correspondiente."""
        disponible = False
        if (self._servicio == servicio and self.isActivo(fechaSolicitud)):
            for i in self._diasLaborales:
                if (i.value - 1 == (fechaSolicitud.weekday())):
                    disponible = True

        return disponible
    

    def isActivo(self, fecha : datetime = datetime.today()) -> bool:
        """Consulta si el contrato esta vigente en la fecha especificada"""
        return self._contrato.consultarVigencia(fecha)

    
    def mostrarInformacion(self) -> str:
        """Devuelve un texto con informacion del empleado dependiendo si tiene contrato
        vigente o no."""
        informacion = ""
        if (self.isActivo()):
            informacion = "Empleado con contrato vigente " + self._nombre + " con número de identificación: " + self._identificacion + ", con cargo: "+ self._cargo
        else:
            informacion = "Empleado sin contrato vigente " + self._nombre + " con número de identificación: " + self._identificacion + ", con cargo: "+ self._cargo
        return informacion

    
    def despedirE(self) -> None:
        """Establece el dia actual como fecha final del contrato para realizar su
        despido"""
        hoy = datetime.today()
        self._contrato.setFechaFin(hoy)
    
    
    def renovarContrato(self, fechaFin : datetime) -> None:
        """Establece una nueva fecha fin del contrato para alargar su vigencia"""
        if (fechaFin > self._contrato.getFechaFin()):
            self._contrato.setFechaFin(fechaFin)

    def contratar(self, contrato : Contrato, cargo : str, servicio : Servicio, diasLaborales : 'list[DiaSemana]'):
        if (not cargo == "x"):
            self.setCargo(cargo)
        
        if (servicio != None):
            self.setServicio(servicio)
        
        if (not diasLaborales.isEmpty()):
            self.setDiasLaborales(diasLaborales)
